From cdabc78fc50a5a1859faebc1e41a3dfbbd007341 Mon Sep 17 00:00:00 2001
From: Armin Novak <armin.novak@thincast.com>
Date: Thu, 9 Jul 2026 08:34:24 +0200
Subject: [PATCH] [core,redirection] check redirection values for validity

Add some basic validity checks for data provided by server.
---
 libfreerdp/core/redirection.c | 25 ++++++++++++++++++++-----
 libfreerdp/core/utils.c       | 33 +++++++++++++++++++++++++++++++++
 libfreerdp/core/utils.h       |  2 ++
 3 files changed, 55 insertions(+), 5 deletions(-)

diff --git a/libfreerdp/core/redirection.c b/libfreerdp/core/redirection.c
index 59c6dbcee..b4b222eb2 100644
--- a/libfreerdp/core/redirection.c
+++ b/libfreerdp/core/redirection.c
@@ -406,6 +406,8 @@ static int rdp_recv_server_redirection_pdu(rdpRdp* rdp, wStream* s)
 	{
 		if (!rdp_redirection_read_unicode_string(s, &(redirection->TargetNetAddress), 80))
 			return -1;
+		if (!utils_is_valid_ip(redirection->TargetNetAddress))
+			return -1;
 	}

 	if (redirection->flags & LB_LOAD_BALANCE_INFO)
@@ -419,13 +421,17 @@ static int rdp_recv_server_redirection_pdu(rdpRdp* rdp, wStream* s)
 		if (!rdp_redirection_read_data(LB_LOAD_BALANCE_INFO, s, &redirection->LoadBalanceInfoLength,
 		                               &redirection->LoadBalanceInfo))
 			return -1;
+		if (!winpr_str_is_valid_urlN((const char*)redirection->LoadBalanceInfo,
+		                             redirection->LoadBalanceInfoLength))
+			return -1;
 	}

 	if (redirection->flags & LB_USERNAME)
 	{
 		if (!rdp_redirection_read_unicode_string(s, &(redirection->Username), 512))
 			return -1;
-
+		if (winpr_str_has_newlines(redirection->Username))
+			return -1;
 		WLog_DBG(TAG, "Username: %s", redirection->Username);
 	}

@@ -433,7 +439,8 @@ static int rdp_recv_server_redirection_pdu(rdpRdp* rdp, wStream* s)
 	{
 		if (!rdp_redirection_read_unicode_string(s, &(redirection->Domain), 52))
 			return -1;
-
+		if (winpr_str_has_newlines(redirection->Domain))
+			return -1;
 		WLog_DBG(TAG, "Domain: %s", redirection->Domain);
 	}

@@ -484,7 +491,8 @@ static int rdp_recv_server_redirection_pdu(rdpRdp* rdp, wStream* s)
 	{
 		if (!rdp_redirection_read_unicode_string(s, &(redirection->TargetFQDN), 512))
 			return -1;
-
+		if (!winpr_str_is_valid_url(redirection->TargetFQDN))
+			return -1;
 		WLog_DBG(TAG, "TargetFQDN: %s", redirection->TargetFQDN);
 	}

@@ -492,7 +500,8 @@ static int rdp_recv_server_redirection_pdu(rdpRdp* rdp, wStream* s)
 	{
 		if (!rdp_redirection_read_unicode_string(s, &(redirection->TargetNetBiosName), 32))
 			return -1;
-
+		if (!winpr_str_is_valid_url(redirection->TargetNetBiosName))
+			return -1;
 		WLog_DBG(TAG, "TargetNetBiosName: %s", redirection->TargetNetBiosName);
 	}

@@ -501,6 +510,8 @@ static int rdp_recv_server_redirection_pdu(rdpRdp* rdp, wStream* s)
 		if (!rdp_redirection_read_data(LB_CLIENT_TSV_URL, s, &redirection->TsvUrlLength,
 		                               &redirection->TsvUrl))
 			return -1;
+		if (!winpr_str_is_valid_urlN((const char*)redirection->TsvUrl, redirection->TsvUrlLength))
+			return -1;
 	}

 	if (redirection->flags & LB_REDIRECTION_GUID)
@@ -541,7 +552,8 @@ static int rdp_recv_server_redirection_pdu(rdpRdp* rdp, wStream* s)
 		{
 			if (!rdp_redirection_read_unicode_string(s, &(redirection->TargetNetAddresses[i]), 80))
 				return -1;
-
+			if (!utils_is_valid_ip(redirection->TargetNetAddresses[i]))
+				return -1;
 			WLog_DBG(TAG, "TargetNetAddresses[%" PRIuz "]: %s", i,
 			         redirection->TargetNetAddresses[i]);
 		}
@@ -569,7 +581,10 @@ int rdp_recv_enhanced_security_redirection_packet(rdpRdp* rdp, wStream* s)
 	status = rdp_recv_server_redirection_pdu(rdp, s);

 	if (status < 0)
+	{
+		WLog_ERR(TAG, "redirection packet invalid, aborting");
 		return status;
+	}

 	if (Stream_GetRemainingLength(s) >= 1)
 	{
diff --git a/libfreerdp/core/utils.c b/libfreerdp/core/utils.c
index 139c1a433..102819915 100644
--- a/libfreerdp/core/utils.c
+++ b/libfreerdp/core/utils.c
@@ -22,6 +22,16 @@
 #include "config.h"
 #endif

+#include <string.h>
+
+#ifdef _WIN32
+#include <ws2tcpip.h>
+#else
+#include <sys/socket.h>
+#include <netinet/in.h>
+#include <arpa/inet.h>
+#endif
+
 #include <winpr/assert.h>

 #include <freerdp/freerdp.h>
@@ -65,3 +75,26 @@ BOOL utils_str_is_empty(const char* str)
 		return TRUE;
 	return FALSE;
 }
+
+static BOOL isValidIPv4(const char* ipAddress)
+{
+	struct sockaddr_in sa;
+	int result;
+	memset(&sa, 0, sizeof(sa));
+	result = inet_pton(AF_INET, ipAddress, &(sa.sin_addr));
+	return result != 0;
+}
+
+static BOOL isValidIPv6(const char* ipAddress)
+{
+	struct sockaddr_in6 sa;
+	int result;
+	memset(&sa, 0, sizeof(sa));
+	result = inet_pton(AF_INET6, ipAddress, &(sa.sin6_addr));
+	return result != 0;
+}
+
+BOOL utils_is_valid_ip(const char* ipAddress)
+{
+	return isValidIPv4(ipAddress) || isValidIPv6(ipAddress);
+}
diff --git a/libfreerdp/core/utils.h b/libfreerdp/core/utils.h
index 432951812..4b150c416 100644
--- a/libfreerdp/core/utils.h
+++ b/libfreerdp/core/utils.h
@@ -30,4 +30,6 @@ BOOL utils_abort_connect(rdpContext* context);
 BOOL utils_str_is_empty(const char* str);
 BOOL utils_str_copy(const char* value, char** dst);

+FREERDP_LOCAL BOOL utils_is_valid_ip(const char* ipAddress);
+
 #endif /* FREERDP_LIB_CORE_UTILS_H */
